File: 202. Happy Number/solution.java
class Solution {
    public boolean isHappy(int n) {
        if(n==1){
            return true;
        }
        
        while (n >= 10) {
            int sum = 0;
            while (n != 0) {
                int d = n % 10;     //个位数
                n = n / 10;         //十位数
                sum += d*d;
            }
            n = sum;
        }
        return n == 1 || n == 7;
    }
}

File: 202. Happy Number/solution.cpp
class Solution {
public:
    int getNextNum(int n){
        int newNum = 0;
        while(n!=0){
            int num = n%10;
            newNum += num*num;
            n = n/10;
        }
        return newNum;
    }

    bool isHappy(int n) {
        int slow = n;
        int fast = getNextNum(n);

        while(fast != 1 && fast != slow){
            slow = getNextNum(slow);
            fast = getNextNum(getNextNum(fast));
        }
        return fast == 1;
        
    }
};
